Default extensions/v1beta1 Deployment's ProgressDeadlineSeconds to MaxInt32

janetkuo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**: Default values should be set in all API versions, because defaulting happens whenever a serialized version is read. When we switched to `apps/v1` as the storage version in `1.10` (#58854), `extensions/v1beta1` `DeploymentSpec.ProgressDeadlineSeconds` gets `apps/v1` default value (`600`) instead of being unset. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #66135 

**Special notes for your reviewer**: We need to cherrypick this fix to 1.10 and 1.11. Note that this fix will only help people who haven't upgraded to 1.10 or 1.11 when the storage version is changed. 

@kubernetes/sig-apps-bugs 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
extensions/v1beta1.Deployment .spec.progressDeadlineSeconds should be defaulted to "no deadline" instead of 600. 
```

janetkuo committed:
Default extensions/v1beta1 Deployment's ProgressDeadlineSeconds to MaxInt32.

1. MaxInt32 has the same meaning as unset, for compatibility
2. Deployment controller treats MaxInt32 the same as unset (nil)

janetkuo committed:
Update API doc of ProgressDeadlineSeconds

enisoc commented on pkg/apis/extensions/validation/validation.go:
```diff
@@ -289,7 +289,8 @@ func ValidateDeploymentSpec(spec *extensions.DeploymentSpec, fldPath *field.Path
 	if spec.ProgressDeadlineSeconds != nil {
 		allErrs = append(allErrs, apivalidation.ValidateNonnegativeField(int64(*spec.ProgressDeadlineSeconds), fldPath.Child("progressDeadlineSeconds"))...)
-		if *spec.ProgressDeadlineSeconds <= spec.MinReadySeconds {
+		if *spec.ProgressDeadlineSeconds > 0 && *spec.ProgressDeadlineSeconds <= spec.MinReadySeconds {
```
Is it safe to relax validation to allow 0? The [API guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#backward-compatibility-gotchas) say, "Validation rules on spec fields can neither be relaxed nor strengthened."

In this case, I'm not too worried about clients being broken by 0, but what happens on the server if, for example, you roll back to a previous patch release after some objects have been persisted with a 0 value? Will the object become inaccessible? Will a "list all" fail? Or does the server handle this gracefully somehow?

janetkuo replied:
Even if the object is still accessible with `0` value after downgrade, the old controller will treat deadline == 0 as "immediate pass deadline" instead of "no deadline". 

It's safer to use max int instead, then.

janetkuo replied:
Updated to use max int32 instead.

smarterclayton commented:
Doesn't this mean that anytime we have an effective default on a pointer
field, where we want to leave the field unset, we can't change the default
across versions?  Pretty sure we've introduced more of these as we've
become quite pointer happy these days.

On Tue, Jul 24, 2018 at 9:12 PM k8s-ci-robot <notifications@github.com>
wrote:

> [APPROVALNOTIFIER] This PR is *NOT APPROVED*
>
> This pull-request has been approved by: *janetkuo
> <https://github.com/kubernetes/kubernetes/pull/66581#>*
> To fully approve this pull request, please assign additional approvers.
> We suggest the following additional approver: *lavalamp*
>
> If they are not already assigned, you can assign the PR to them by writing /assign
> @lavalamp in a comment when ready.
>
> The full list of commands accepted by this bot can be found here
> <https://go.k8s.io/bot-commands>.
>
> The pull request process is described here
> <https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process>
> Needs approval from an approver in each of these files:
>
>    - *api/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/api/OWNERS>*
>    - *pkg/apis/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/OWNERS>*
>    - pkg/controller/deployment/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/OWNERS>
>    [janetkuo]
>    - *staging/src/k8s.io/api/OWNERS
>    <https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/OWNERS>*
>
> Approvers can indicate their approval by writing /approve in a comment
> Approvers can cancel approval by writing /approve cancel in a comment
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/66581#issuecomment-407600663>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_px-Isf0NgpgiMyDuz2lAHyBlD8jxks5uJ8XkgaJpZM4Veh72>
> .
>

janetkuo commented:
> Doesn't this mean that anytime we have an effective default on a pointer field, where we want to leave the field unset, we can't change the default across versions?  

@smarterclayton Yes. We cannot leave a field unset in one version but defaulted in another version, because defaulting are not just applied when an object is created, but every time a serialized version is read. 

> Pretty sure we've introduced more of these as we've become quite pointer happy these days.

I opened an umbrella issue #66582 to track other violations.

janetkuo commented:
/test pull-kubernetes-e2e-gce

enisoc commented on pkg/controller/deployment/progress.go:
```diff
@@ -47,7 +47,7 @@ func (dc *DeploymentController) syncRolloutStatus(allRSs []*apps.ReplicaSet, new
 	// Check for progress only if there is a progress deadline set and the latest rollout
 	// hasn't completed yet.
-	if d.Spec.ProgressDeadlineSeconds != nil && !isCompleteDeployment {
+	if !util.NoProgressDeadline(d) && !isCompleteDeployment {
```
Nit: could we invert the function to `HasProgressDeadline` to avoid the double negative?

janetkuo replied:
Good point. Done.

enisoc commented on pkg/apis/extensions/types.go:
```diff
@@ -127,8 +127,8 @@ type DeploymentSpec struct {
-	// not be estimated during the time a deployment is paused. This is not set
-	// by default.
+	// not be estimated during the time a deployment is paused. This is set to
+	// the max value of int32 (1<<31 - 1) by default, which means "no deadline".
```
Nit: Since this may end up in generated user docs, it might be better to give the explicit value instead (2147483647) so it looks the same way people will see it printed in objects, and so people don't need to be familiar with bit shift syntax.

janetkuo replied:
Done.

enisoc reviewed (approved):
LGTM other than nits.

janetkuo committed:
Autogen

1. hack/update-generated-protobuf.sh
2. hack/update-generated-swagger-docs.sh
3. hack/update-swagger-spec.sh
4. hack/update-openapi-spec.sh
5. hack/update-api-reference-docs.sh

janetkuo commented:
@enisoc PTAL

enisoc reviewed (approved):
/lgtm

janetkuo commented:
/assign @smarterclayton 

Would you approve it? Thanks!

tnozicka reviewed (commented):
nit; lgtm otherwise

smarterclayton commented:
/approve

janetkuo commented on pkg/controller/deployment/progress_test.go:
```diff
@@ -330,7 +338,7 @@ func TestSyncRolloutStatus(t *testing.T) {
 			switch {
 			case newCond == nil:
-				if test.d.Spec.ProgressDeadlineSeconds != nil {
+				if test.d.Spec.ProgressDeadlineSeconds != nil && *test.d.Spec.ProgressDeadlineSeconds != math.MaxInt32 {
```
Intentionally write it explicitly in the test, so that if we break `deploymentutil.HasProgressDeadline()` in the future, this test will still catch that.

nikhita commented:
This should include a release note.

@janetkuo Can you update the main PR body to include a release note? Thanks!

janetkuo commented:
Added a release note. Thanks!
